### PR TITLE
ci: Enable FC CI for kata 2.x

### LINKED ISSUE #3076: enable FC CI for kata 2.x

### PR DESCRIPTION
**Pull request by GabyCT**
This PR enables FC CI for kata 2.x

Fixes #3076

Signed-off-by: Gabriela Cervantes <gabriela.cervantes.tellez@intel.com>

### CODE REVIEW

**Comment by GabyCT** [comment]
@chavafg @amshinde  I tried to enable `devicemapper` with containerd+FC but I still see failures in the tests. I am not sure if I am missing something else, the configuration of containerd is the following
```
[debug]
  level = "debug"
[plugins]
  [plugins.devmapper]
    pool_name = "contd-thin-pool"
    base_image_size = "4096MB"
  [plugins.cri]
    [plugins.cri.containerd]
      [plugins.cri.containerd.runtimes]
        [plugins.cri.containerd.runtimes.runc]
           runtime_type = "io.containerd.runc.v1"
           [plugins.cri.containerd.runtimes.runc.options]
             BinaryName = "/usr/local/sbin/runc"
             Root = ""
        [plugins.cri.containerd.runtimes.kata]
           runtime_type = "io.containerd.kata.v2"
           privileged_without_host_devices = true
```
And the test is using the following configuration
```
 [plugins]
 [plugins.devmapper]
     pool_name = "contd-thin-pool"
     base_image_size = "4096MB"
   [plugins.cri]
     [plugins.cri.containerd]
 	default_runtime_name = "io-containerd-kata-v2"
       [plugins.cri.containerd.runtimes.io-containerd-kata-v2]
         runtime_type = "io.containerd.kata.v2"
         [plugins.cri.containerd.runtimes.io-containerd-kata-v2.options]
           Runtime = "io.containerd.kata.v2"
 [plugins.linux]
        shim = "/usr/local/bin/containerd-shim"
```
## Kata Containerd Shim v2 logs

<details>
<summary>Kata Containerd Shim v2</summary>
<p>

Recent  problems found in system journal:
```
time="2020-11-23T23:53:04.720632619Z" level=error msg="getting vm status failed" error="Get \"http://localhost/\": dial unix /run/vc/firecracker/f3a5c7a5069b3f1aee6e98f5366ebb92/root/run/firecracker.socket: connect: no such file or directory" pid=30123 sandbox=f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:04.737523442Z" level=error msg="getting vm status failed" error="Get \"http://localhost/\": dial unix /run/vc/firecracker/64350b7f34d8301ea4511301e335cfc4/root/run/firecracker.socket: connect: no such file or directory" pid=30147 sandbox=64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:04.770473181Z" level=error msg="getting vm status failed" error="Get \"http://localhost/\": dial unix /run/vc/firecracker/7707a8398fe085e6abf9909cdec14cfe/root/run/firecracker.socket: connect: no such file or directory" pid=30124 sandbox=7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:06.00965037Z" level=error msg="getting vm status failed" error="Get \"http://localhost/\": dial unix /run/vc/firecracker/ac8560812168208b0a833366e15ffa61/root/run/firecracker.socket: connect: no such file or directory" pid=30234 sandbox=ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227 source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:06.025578589Z" level=error msg="createContainer failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c/rootfs\" to \"/run/kata-containers/053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c/rootfs\", with error: No such file or directory (os error 2)" pid=30121 sandbox=053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c source=virtcontainers subsystem=kata_agent
time="2020-11-23T23:53:06.025984692Z" level=error msg="container create failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c/rootfs\" to \"/run/kata-containers/053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c/rootfs\", with error: No such file or directory (os error 2)" pid=30121 sandbox=053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c source=virtcontainers subsystem=container
time="2020-11-23T23:53:06.026248494Z" level=warning error="no such file or directory" pid=30121 rootfs-dir=/run/kata-containers/shared/sandboxes/053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c/mounts/053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c/rootfs sandbox=053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c source=virtcontainers subsystem=mount
time="2020-11-23T23:53:06.034950959Z" level=error msg="createContainer failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d/rootfs\" to \"/run/kata-containers/64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d/rootfs\", with error: No such file or directory (os error 2)" pid=30147 sandbox=64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d source=virtcontainers subsystem=kata_agent
time="2020-11-23T23:53:06.035296761Z" level=error msg="container create failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d/rootfs\" to \"/run/kata-containers/64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d/rootfs\", with error: No such file or directory (os error 2)" pid=30147 sandbox=64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d source=virtcontainers subsystem=container
time="2020-11-23T23:53:06.035396362Z" level=warning error="no such file or directory" pid=30147 rootfs-dir=/run/kata-containers/shared/sandboxes/64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d/mounts/64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d/rootfs sandbox=64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d source=virtcontainers subsystem=mount
time="2020-11-23T23:53:06.051851985Z" level=error msg="createContainer failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce/rootfs\" to \"/run/kata-containers/f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce/rootfs\", with error: No such file or directory (os error 2)" pid=30123 sandbox=f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce source=virtcontainers subsystem=kata_agent
time="2020-11-23T23:53:06.052064086Z" level=error msg="container create failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce/rootfs\" to \"/run/kata-containers/f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce/rootfs\", with error: No such file or directory (os error 2)" pid=30123 sandbox=f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce source=virtcontainers subsystem=container
time="2020-11-23T23:53:06.052188287Z" level=warning error="no such file or directory" pid=30123 rootfs-dir=/run/kata-containers/shared/sandboxes/f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce/mounts/f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce/rootfs sandbox=f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce source=virtcontainers subsystem=mount
time="2020-11-23T23:53:06.103164467Z" level=error msg="createContainer failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca/rootfs\" to \"/run/kata-containers/7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca/rootfs\", with error: No such file or directory (os error 2)" pid=30124 sandbox=7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca source=virtcontainers subsystem=kata_agent
time="2020-11-23T23:53:06.108165905Z" level=error msg="container create failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca/rootfs\" to \"/run/kata-containers/7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca/rootfs\", with error: No such file or directory (os error 2)" pid=30124 sandbox=7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca source=virtcontainers subsystem=container
time="2020-11-23T23:53:06.108315806Z" level=warning error="no such file or directory" pid=30124 rootfs-dir=/run/kata-containers/shared/sandboxes/7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca/mounts/7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca/rootfs sandbox=7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca source=virtcontainers subsystem=mount
time="2020-11-23T23:53:06.798702053Z" level=error msg="createContainer failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227/rootfs\" to \"/run/kata-containers/ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227/rootfs\", with error: No such file or directory (os error 2)" pid=30234 sandbox=ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227 source=virtcontainers subsystem=kata_agent
time="2020-11-23T23:53:06.799090956Z" level=error msg="container create failed" error="rpc error: code = Internal desc = failed to mount \"/run/kata-containers/shared/containers/ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227/rootfs\" to \"/run/kata-containers/ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227/rootfs\", with error: No such file or directory (os error 2)" pid=30234 sandbox=ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227 source=virtcontainers subsystem=container
time="2020-11-23T23:53:06.799333858Z" level=warning error="no such file or directory" pid=30234 rootfs-dir=/run/kata-containers/shared/sandboxes/ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227/mounts/ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227/rootfs sandbox=ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227 source=virtcontainers subsystem=mount
time="2020-11-23T23:53:21.042793162Z" level=warning msg="VM still running after waiting 15s" pid=30147 sandbox=64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:21.043143465Z" level=warning msg="sandbox cgroups path is empty" pid=30147 sandbox=64350b7f34d8301ea4511301e335cfc45737dc40149485518d3f08e8b431510d source=virtcontainers subsystem=sandbox
time="2020-11-23T23:53:21.081675755Z" level=warning msg="VM still running after waiting 15s" pid=30121 sandbox=053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:21.081909257Z" level=warning msg="sandbox cgroups path is empty" pid=30121 sandbox=053bc5206cd0fe3338de80c591f69b35bd460d6c093f348935319920c72a5f1c source=virtcontainers subsystem=sandbox
time="2020-11-23T23:53:21.105420934Z" level=warning msg="VM still running after waiting 15s" pid=30123 sandbox=f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:21.105597135Z" level=warning msg="sandbox cgroups path is empty" pid=30123 sandbox=f3a5c7a5069b3f1aee6e98f5366ebb92c8c1c74cc3cc52596d7ef55890df61ce source=virtcontainers subsystem=sandbox
time="2020-11-23T23:53:21.155667312Z" level=warning msg="VM still running after waiting 15s" pid=30124 sandbox=7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:21.155878113Z" level=warning msg="sandbox cgroups path is empty" pid=30124 sandbox=7707a8398fe085e6abf9909cdec14cfe1fa3a85c05c8d9e470e4a83bb33b03ca source=virtcontainers subsystem=sandbox
time="2020-11-23T23:53:21.80637431Z" level=warning msg="VM still running after waiting 15s" pid=30234 sandbox=ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227 source=virtcontainers subsystem=firecracker
time="2020-11-23T23:53:21.806685812Z" level=warning msg="sandbox cgroups path is empty" pid=30234 sandbox=ac8560812168208b0a833366e15ffa61b0f7d1afac2ee4869f07119d76a42227 source=virtcontainers subsystem=sandbox
```

</p>
</details>

**Comment by liubin** [comment]
Hi @GabyCT I'm using these configs and it works. Help it will be helpful.

```
[plugins]
  [plugins."io.containerd.grpc.v1.cri"]
    [plugins."io.containerd.grpc.v1.cri".containerd]
      snapshotter = "devmapper"
  [plugins."io.containerd.snapshotter.v1.devmapper"]
    pool_name = "devpool"
    root_path = "/var/lib/containerd/devmapper"
    base_image_size = "10GB"
```

File: .ci/containerd_devmapper_setup.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+[Unit]
+Description=Setup containerd devmapper device
+DefaultDependencies=no
+After=systemd-udev-settle.service
+Before=lvm2-activation-early.service
+Wants=systemd-udev-settle.service
+[Service]
+Type=oneshot
+RemainAfterExit=true
+ExecStart=-/sbin/losetup /dev/loop20 /var/lib/containerd/devmapper/data-disk.img
```
Comment (liubin) [inline .ci/containerd_devmapper_setup.sh:36]: Don't assure that whether the fixed device name will lead errors, but I suggest borrow parts of https://github.com/containerd/containerd/tree/master/snapshots/devmapper

File: .ci/run.sh [hunk 2]
```diff
@@ -49,6 +49,10 @@ case "${CI_JOB}" in
 		echo "INFO: Running complete e2e kubernetes tests"
 		sudo -E PATH="$PATH" CRI_RUNTIME="containerd" bash -c "make kubernetes-e2e"
 		;;
+	"FIRECRACKER")
+		echo "INFO: Containerd checks"
```
Comment (liubin) [inline .ci/run.sh:53]: In these `case` branches, there are 3 `echo "INFO: Containerd checks"`, I'd suggest echo more distinct info by adding the case value(clh or firecracker or qemu) into it.

**Comment by GabyCT** [comment]
> Hi @GabyCT I'm using these configs and it works. Help it will be helpful.
> 
> ```
> [plugins]
>   [plugins."io.containerd.grpc.v1.cri"]
>     [plugins."io.containerd.grpc.v1.cri".containerd]
>       snapshotter = "devmapper"
>   [plugins."io.containerd.snapshotter.v1.devmapper"]
>     pool_name = "devpool"
>     root_path = "/var/lib/containerd/devmapper"
>     base_image_size = "10GB"
> ```
> 
> I checked the log and couldn't find the `snapshotter = "devmapper"`, is this the cause?

@liubin thanks for the feedback, one question, the `snapshotter="devmapper"` needs to be added in this place?
```
 [plugins]
 [plugins.devmapper]
     pool_name = "contd-thin-pool"
     base_image_size = "4096MB"
   [plugins.cri]
     [plugins.cri.containerd]
 	default_runtime_name = "io-containerd-kata-v2"
       [plugins.cri.containerd.runtimes.io-containerd-kata-v2]
         runtime_type = "io.containerd.kata.v2"
         **snapshotter = "devmapper"**
         [plugins.cri.containerd.runtimes.io-containerd-kata-v2.options]
           Runtime = "io.containerd.kata.v2"
 [plugins.linux]
        shim = "/usr/local/bin/containerd-shim"
```
thanks

**Comment by liubin** [comment]
@GabyCT I think it's better to set `snapshotter = "devmapper"`, at least it's not harmful.

File: .ci/containerd_devmapper_setup.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KATA_HYPERVISOR="${KATA_HYPERVISOR:-qemu}"
+
+if [ "$KATA_HYPERVISOR" != "firecracker" ]; then
+	echo "Devicemapper configuration is only for Firecracker"
+	exit
```
Comment (jodh-intel) [inline .ci/containerd_devmapper_setup.sh:15]: Can you make this clearer (`echo "ERROR: ..."`, `echo "WARNING: ..."` or `echo "INFO: ..."`?) and also please specify an return code value to `exit`.

The same applies to other echo messages below.

**Comment by jodh-intel** [comment]
@GabyCT - also this is now conflicted.

**Comment by fidencio** [comment]
I'm really sorry for the disruption but in order to do the branch name changes announced [here](http://lists.katacontainers.io/pipermail/kata-dev/2021-January/001695.html) I'm adding the "do-not-merge" and the "needs-branch-adjustment" labels to the PR.

I'll add a new comment once the branch changes are done and once it happens I'd like to ask you to resubmit the PR against the new `main` branch, possibly linking to this PR in case you find the conversations / reviews already made important to proceed.

**Comment by fidencio** [comment]
Kata Containers repos are back to business.  Thanks a lot for your patience and I'm very happy to announce that there's no changes needed from your side as renaming the branches automagically reflected here.

If you notice something weird, some missing CI, please, do not hesitate on contacting us here or on #kata-dev on bit.ly/katacontainersslack.

Also, this is a good opportunity to have your PR rebased atop of `main`, to ensure your PR follow all the guidelines (https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format) enforced by this community, or even to re-evaluate whether the PR is currently needed.

Last but not least,  as the default branch has been renamed to `main`, if you have a local clone you'd need to update it and you do so by running:
```
git branch -m 2.0-dev main
git fetch origin
git branch -u origin/main main
```

Happy hacking!

File: .ci/containerd_devmapper_setup.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+#
+# Copyright (c) 2020 Intel Corporation
```
Comment (devimc) [inline .ci/containerd_devmapper_setup.sh:3]: 2021

**Review by devimc — APPROVED** [review]
thanks @GabyCT - one nit

**Review by jcvenegas — APPROVED** [review]
thanks Gaby, at some point would  be nice to have a list defined somewhere to know the list of items we skip for fc.

**Comment by GabyCT** [comment]
@jcvenegas yes I will raise an issue with all the tests that are failing

File: .ci/containerd_devmapper_setup.sh [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+Description=Setup containerd devmapper device
+DefaultDependencies=no
+After=systemd-udev-settle.service
+Before=lvm2-activation-early.service
+Wants=systemd-udev-settle.service
+[Service]
+Type=oneshot
+RemainAfterExit=true
+ExecStart=-/sbin/losetup /dev/loop20 /var/lib/containerd/devmapper/data-disk.img
+ExecStart=-/sbin/losetup /dev/loop21 /var/lib/containerd/devmapper/meta-disk.img
```
Comment (jcvenegas) [inline .ci/containerd_devmapper_setup.sh:41]: Second nit, would be good to not use fixed names

**Pull request merged**